subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,574 @@
+
+	in->offset += len;
+
+	return len;
+}
+
+static size_t get_opaque(struct lwm2m_input_context *in,
+			 uint8_t *value, size_t buflen,
+			 struct lwm2m_opaque_context *opaque,
+			 bool *last_block)
+{
+	cbor_state_t states[1];
+	cbor_string_type_t info = {0};
+	int ret;
+
+	/* Get the CBOR header only on first read. */
+	if (opaque->remaining == 0) {
+
+		/* FIXME: When data is spread between non-contiguous blocks this range setting will
+		 * break the whole thing. Luckily we get the data length out though.
+		 */
+		new_state(states, 1, ICTX_BUF_R_REGION(in), 1);
+
+		ret = bstrx_cbor_start_decode(states, &info);
+
+		/* FIXME: Uncomment once the range check is working */
+		(void)ret;
+		/*
+		 * if(!ret) {
+		 *	LOG_WRN("unable to decode opaque data header");
+		 *	return -EBADMSG;
+		 * }
+		 */
+
+		opaque->len = info.len;
+		opaque->remaining = info.len;
+
+		/* FIXME: Workaround for the bad range check.
+		 * Now it's CBOR data item header plus possible extended count
+		 */
+		if (info.len == 0) {
+			LOG_WRN("unable to decode opaque data header");
+			return -EBADMSG;
+		} else if (info.len <= 23) {
+			in->offset += 1;
+		} else if (info.len <= UINT8_MAX) {
+			in->offset += 1 + 1;
+		} else if (info.len <= UINT16_MAX) {
+			in->offset += 1 + 2;
+		} else if (info.len <= UINT32_MAX) {
+			in->offset += 1 + 4;
+		}
+	}
+
+	return lwm2m_engine_get_opaque_more(in, value, buflen,
+					    opaque, last_block);
+}
+
+static size_t get_objlnk(struct lwm2m_input_context *in,
+			 struct lwm2m_objlnk *value)
+{
+	char objlnk[sizeof("65535:65535")] = {0};
+
+	value->obj_id = LWM2M_OBJLNK_NULL;
+	value->obj_inst = LWM2M_OBJLNK_NULL;
+
+	int len = get_string(in, objlnk, sizeof(objlnk)-1);
+
+	char *end;
+	char *idp = objlnk;
+
+	errno = 0;
+
+	for (int idx = 0; idx < 2; idx++) {
+		unsigned long id = strtoul(idp, &end, 10);
+
+		idp = end;
+
+		if (errno == ERANGE || id > 65535) {
+			LOG_WRN("decoded id %lu out of range[0..65535]", id);
+			errno = 0;
+			return -EBADMSG;
+		}
+
+		switch (idx) {
+		case 0:
+			value->obj_id = id;
+			continue;
+		case 1:
+			value->obj_inst = id;
+			continue;
+		}
+	}
+
+	if (value->obj_inst != LWM2M_OBJLNK_NULL && (value->obj_inst == LWM2M_OBJLNK_NULL)) {
+		LOG_WRN("decoded obj inst id without obj id");
+		return -EBADMSG;
+	}
+
+	return len;
+}
+
+const struct lwm2m_writer cbor_writer = {
+	.put_s8 = put_s8,
+	.put_s16 = put_s16,
+	.put_s32 = put_s32,
+	.put_s64 = put_s64,
+	.put_string = put_string,
+	.put_float = NULL,
+	.put_time = put_time,
+	.put_bool = put_bool,
+	.put_opaque = put_opaque,
+	.put_objlnk = put_objlnk,
+};
+
+const struct lwm2m_reader cbor_reader = {
+	.get_s32 = get_s32,
+	.get_s64 = get_s64,
+	.get_time = get_time,
+	.get_string = get_string,
+	.get_float = NULL,
+	.get_bool = get_bool,
+	.get_opaque = get_opaque,
+	.get_objlnk = get_objlnk,
+};
+
+int do_read_op_cbor(struct lwm2m_message *msg)
+{
+	/* Can only return single resource */
+	if (msg->path.level < 3U) {
```

Comment:
Would be good to use symbols for path levels, i. e. `LWM2M_PATH_LEVEL_RESOURCE` etc.